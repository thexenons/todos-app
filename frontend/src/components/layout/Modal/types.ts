import type { PropsWithChildren } from "react";

import { BREAKPOINT } from "../../../theme/common/breakpoints";

export interface ModalProps extends PropsWithChildren {
	isOpen: boolean;
	onClose: () => void;
	onCancel?: () => void;
	onAccept?: () => void;
	wrapperId?: string;
	maxWidth?: BREAKPOINT;
	fullWidth?: boolean;
}
