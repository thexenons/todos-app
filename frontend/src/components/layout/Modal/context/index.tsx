import { createContext, useContext } from "react";

import { ModalProps } from "../types";

type ModalContextProps = Pick<
	ModalProps,
	"isOpen" | "onClose" | "onCancel" | "onAccept"
>;

const ModalContext = createContext<ModalContextProps>({
	isOpen: false,
	onClose: () => null,
});

export const ModalContextProvider = ModalContext.Provider;

export const useModalContext = () => useContext(ModalContext);
