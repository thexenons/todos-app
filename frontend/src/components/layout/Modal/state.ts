import { atom } from "jotai";

export const openedModalsAtom = atom<number>(0);
