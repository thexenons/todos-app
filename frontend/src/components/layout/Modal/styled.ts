import styled, { css } from "styled-components";

import { BREAKPOINT } from "../../../theme/common/breakpoints";
import Button from "../../atoms/Button";
import { ModalProps } from "./types";

export const ModalBackdrop = styled.div`
	position: fixed;
	top: 0;
	left: 0;
	width: 100%;
	height: 100%;
	background-color: rgba(0, 0, 0, 0.5);
	z-index: ${({ theme }) => theme.zIndex.modal};
`;

export const ModalWrapper = styled.div<{
	$maxWidth: ModalProps["maxWidth"];
	$fullWidth: ModalProps["fullWidth"];
}>`
	position: fixed;
	top: 50%;
	left: 50%;
	transform: translate(-50%, -50%);
	z-index: ${({ theme }) => theme.zIndex.modal};
	color: ${({ theme }) => theme.colors.text.primary};
	background-color: ${({ theme }) => theme.colors.background.paper};
	${({ theme }) => theme.mixins.flex({ direction: "column" })}
	box-shadow: ${({ theme }) =>
		`${theme.spacing(0.5)} ${theme.spacing(1)} ${theme.spacing(
			3
		)} rgba(0,0,0,0.4)`};

	${({ $fullWidth }) =>
		$fullWidth &&
		css`
			width: 100%;
			height: 100%;
		`}

	max-width: ${({ theme, $maxWidth = BREAKPOINT.xl }) =>
		`${theme.breakpoints[$maxWidth]}px`};

	${({ theme }) => theme.media.md.up} {
		min-width: 300px;
		height: auto;
		width: ${({ theme, $fullWidth }) =>
			$fullWidth ? `calc(100% - ${theme.spacing(4)})` : "auto"};
		border-radius: ${({ theme }) => theme.spacing(2)};
	}
`;

export const ModalContent = styled.div`
	${({ theme }) => theme.mixins.padding({ all: 2 })}
	flex: 1 1 auto;
`;

export const ModalHeader = styled.div`
	${({ theme }) =>
		theme.mixins.flex({
			align: "center",
			justify: "space-between",
			gap: theme.spacing(2),
		})}
	${({ theme }) => theme.mixins.padding({ x: 2, top: 2 })}
`;

export const ModalTitle = styled.h6`
	${({ theme }) => theme.mixins.typography("h6")}
	margin: 0;
`;

export const ModalCloseButton = styled(Button)`
	margin-left: auto;
`;

export const ModalFooter = styled.div`
	${({ theme }) =>
		theme.mixins.flex({
			align: "center",
			justify: "center",
			gap: theme.spacing(2),
		})}
	${({ theme }) => theme.mixins.padding({ x: 2, bottom: 2 })}
`;
