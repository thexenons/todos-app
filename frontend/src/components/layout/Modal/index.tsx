import { ComponentProps, FC, useCallback } from "react";
import { createPortal } from "react-dom";
import { MdClose } from "react-icons/md";

import Button from "../../atoms/Button";
import AcceptButton from "../../molecules/AcceptButton";
import CancelButton from "../../molecules/CancelButton";
import { ModalContextProvider, useModalContext } from "./context";
import useModal from "./hooks/useModal";
import * as S from "./styled";
import { ModalProps } from "./types";

const ModalView: FC<ModalProps> = ({
	isOpen,
	onClose,
	onCancel,
	onAccept,
	wrapperId = "react-portal-wrapper",
	maxWidth,
	fullWidth,
	children,
	...rest
}) => {
	const { wrapperElement } = useModal({ wrapperId, isOpen });

	if (wrapperElement === null || !isOpen) return null;

	return createPortal(
		<ModalContextProvider value={{ isOpen, onClose, onCancel, onAccept }}>
			<S.ModalBackdrop aria-hidden="true" onClick={onClose} />
			<S.ModalWrapper $fullWidth={fullWidth} $maxWidth={maxWidth} {...rest}>
				{children}
			</S.ModalWrapper>
		</ModalContextProvider>,
		wrapperElement
	);
};

const ModalCloseButton: FC<Omit<ComponentProps<typeof Button>, "children">> = (
	props
) => {
	const { onClose } = useModalContext();

	return (
		<S.ModalCloseButton
			variant={Button.variants.contained}
			colorVariant={Button.colorVariants.primary}
			isRounded
			onClick={onClose}
			{...props}
		>
			<Button.Icon>
				<MdClose />
			</Button.Icon>
		</S.ModalCloseButton>
	);
};

const ModalCancelButton: typeof CancelButton = (props) => {
	const { onClose, onCancel } = useModalContext();

	const onClick = useCallback(() => {
		onClose?.();
		onCancel?.();
	}, [onCancel, onClose]);

	return <CancelButton onClick={onClick} {...props} />;
};
const ModalAcceptButton: typeof AcceptButton = (props) => {
	const { onClose, onAccept } = useModalContext();

	const onClick = useCallback(() => {
		onClose?.();
		onAccept?.();
	}, [onAccept, onClose]);

	return <AcceptButton onClick={onClick} {...props} />;
};

const Modal = ModalView as typeof ModalView & {
	Header: typeof S.ModalHeader;
	Title: typeof S.ModalTitle;
	CloseButton: typeof ModalCloseButton;
	CancelButton: typeof ModalCancelButton;
	AcceptButton: typeof ModalAcceptButton;
	Content: typeof S.ModalContent;
	Footer: typeof S.ModalFooter;
};
Modal.Header = S.ModalHeader;
Modal.Title = S.ModalTitle;
Modal.CloseButton = ModalCloseButton;
Modal.CancelButton = ModalCancelButton;
Modal.AcceptButton = ModalAcceptButton;
Modal.Content = S.ModalContent;
Modal.Footer = S.ModalFooter;

export default Modal;
