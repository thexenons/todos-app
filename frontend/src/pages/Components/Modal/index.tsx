import { FC, useCallback, useState } from "react";

import Button from "../../../components/atoms/Button";
import Modal from "../../../components/layout/Modal";
import { ModalProps } from "../../../components/layout/Modal/types";
import { BREAKPOINT } from "../../../theme/common/breakpoints";
import * as SS from "../styled";

const ModalComponentsPage: FC = () => {
	const [isOpen, setOpen] = useState<ModalProps["isOpen"]>(false);
	const [isAnotherOpen, setAnotherOpen] = useState<ModalProps["isOpen"]>(false);
	const [fullWidth, setFullWidth] = useState<ModalProps["fullWidth"]>();
	const [maxWidth, setMaxWidth] = useState<ModalProps["maxWidth"]>();

	const [enableHeader, setEnableHeader] = useState(true);
	const [enableTitle, setEnableTitle] = useState(true);
	const [enableCloseButton, setEnableCloseButton] = useState(true);
	const [enableFooter, setEnableFooter] = useState(true);
	const [enableAcceptButton, setEnableAcceptButton] = useState(true);
	const [enableCancelButton, setEnableCancelButton] = useState(true);

	const openModal = useCallback(() => setOpen(true), []);
	const closeModal = useCallback(() => setOpen(false), []);
	const openAnotherModal = useCallback(() => setAnotherOpen(true), []);
	const closeAnotherModal = useCallback(() => setAnotherOpen(false), []);

	return (
		<SS.ComponentsListWrapper>
			<h1>Modal</h1>
			<SS.ComponentsListWrapper>
				<h2>Options</h2>
				<SS.ComponentsRowWrapper>
					<p>fullWidth: {fullWidth ? "Yes" : "No"}</p>
					<Button onClick={() => setFullWidth(!fullWidth)}>
						Toggle fullWidth
					</Button>
				</SS.ComponentsRowWrapper>
				<SS.ComponentsRowWrapper>
					<p>maxWidth: {maxWidth || "No"}</p>
					<Button onClick={() => setMaxWidth(undefined)}>Set No</Button>
					<Button onClick={() => setMaxWidth(BREAKPOINT.sm)}>Set sm</Button>
					<Button onClick={() => setMaxWidth(BREAKPOINT.md)}>Set md</Button>
					<Button onClick={() => setMaxWidth(BREAKPOINT.lg)}>Set lg</Button>
					<Button onClick={() => setMaxWidth(BREAKPOINT.xl)}>Set xl</Button>
				</SS.ComponentsRowWrapper>
				<SS.ComponentsRowWrapper>
					<p>Header: {enableHeader ? "Yes" : "No"}</p>
					<Button onClick={() => setEnableHeader(!enableHeader)}>
						Toggle Header
					</Button>
				</SS.ComponentsRowWrapper>
				{enableHeader && (
					<>
						<SS.ComponentsRowWrapper>
							<p>Title: {enableTitle ? "Yes" : "No"}</p>
							<Button onClick={() => setEnableTitle(!enableTitle)}>
								Toggle Title
							</Button>
						</SS.ComponentsRowWrapper>
						<SS.ComponentsRowWrapper>
							<p>CloseButton: {enableCloseButton ? "Yes" : "No"}</p>
							<Button onClick={() => setEnableCloseButton(!enableCloseButton)}>
								Toggle CloseButton
							</Button>
						</SS.ComponentsRowWrapper>
					</>
				)}
				<SS.ComponentsRowWrapper>
					<p>Footer: {enableFooter ? "Yes" : "No"}</p>
					<Button onClick={() => setEnableFooter(!enableFooter)}>
						Toggle Footer
					</Button>
				</SS.ComponentsRowWrapper>
				{enableFooter && (
					<>
						<SS.ComponentsRowWrapper>
							<p>AcceptButton: {enableAcceptButton ? "Yes" : "No"}</p>
							<Button
								onClick={() => setEnableAcceptButton(!enableAcceptButton)}
							>
								Toggle AcceptButton
							</Button>
						</SS.ComponentsRowWrapper>
						<SS.ComponentsRowWrapper>
							<p>CancelButton: {enableCancelButton ? "Yes" : "No"}</p>
							<Button
								onClick={() => setEnableCancelButton(!enableCancelButton)}
							>
								Toggle CancelButton
							</Button>
						</SS.ComponentsRowWrapper>
					</>
				)}
			</SS.ComponentsListWrapper>
			<SS.ComponentsListWrapper>
				<h2>Actions</h2>
				<div>
					<Button onClick={openModal}>Open modal</Button>
				</div>
			</SS.ComponentsListWrapper>
			<Modal
				isOpen={isOpen}
				onClose={closeModal}
				fullWidth={fullWidth}
				maxWidth={maxWidth}
			>
				{enableHeader && (
					<Modal.Header>
						{enableTitle && <Modal.Title>Modal title</Modal.Title>}
						{enableCloseButton && <Modal.CloseButton />}
					</Modal.Header>
				)}
				<Modal.Content>
					<p>
						Lorem, ipsum dolor sit amet consectetur adipisicing elit. Obcaecati
						deleniti maiores expedita nemo placeat animi voluptates architecto
						corrupti consequuntur molestiae perspiciatis, ut atque, nisi velit
						ad eveniet impedit sapiente dignissimos!
					</p>
					<p>
						Lorem, ipsum dolor sit amet consectetur adipisicing elit. Obcaecati
						deleniti maiores expedita nemo placeat animi voluptates architecto
						corrupti consequuntur molestiae perspiciatis, ut atque, nisi velit
						ad eveniet impedit sapiente dignissimos!
					</p>
					<p>
						Lorem, ipsum dolor sit amet consectetur adipisicing elit. Obcaecati
						deleniti maiores expedita nemo placeat animi voluptates architecto
						corrupti consequuntur molestiae perspiciatis, ut atque, nisi velit
						ad eveniet impedit sapiente dignissimos!
					</p>
					<p>
						Lorem, ipsum dolor sit amet consectetur adipisicing elit. Obcaecati
						deleniti maiores expedita nemo placeat animi voluptates architecto
						corrupti consequuntur molestiae perspiciatis, ut atque, nisi velit
						ad eveniet impedit sapiente dignissimos!
					</p>
					<Button onClick={openAnotherModal}>Open another modal</Button>
				</Modal.Content>
				{enableFooter && (
					<Modal.Footer>
						{enableCancelButton && <Modal.CancelButton />}
						{enableAcceptButton && <Modal.AcceptButton />}
					</Modal.Footer>
				)}
			</Modal>
			<Modal
				isOpen={isAnotherOpen}
				onClose={closeAnotherModal}
				fullWidth={fullWidth}
				maxWidth={maxWidth}
			>
				{enableHeader && (
					<Modal.Header>
						{enableTitle && <Modal.Title>Modal title</Modal.Title>}
						{enableCloseButton && <Modal.CloseButton />}
					</Modal.Header>
				)}
				<Modal.Content>
					<p>Another Modal</p>
				</Modal.Content>
				{enableFooter && (
					<Modal.Footer>
						{enableCancelButton && <Modal.CancelButton />}
						{enableAcceptButton && <Modal.AcceptButton />}
					</Modal.Footer>
				)}
			</Modal>
		</SS.ComponentsListWrapper>
	);
};

export default ModalComponentsPage;
