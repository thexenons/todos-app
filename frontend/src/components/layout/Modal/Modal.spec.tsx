import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { vi } from "vitest";

import BaseTestComponent from "../../../tests/BaseTestComponent";
import Modal from ".";

describe("Modal", () => {
	it("Render", () => {
		const { container } = render(
			<BaseTestComponent>
				<Modal isOpen onClose={() => null} />
			</BaseTestComponent>
		);
		expect(container).toMatchSnapshot();
	});

	it("Render with fullWidth", () => {
		const { container } = render(
			<BaseTestComponent>
				<Modal isOpen onClose={() => null} fullWidth />
			</BaseTestComponent>
		);
		expect(container).toMatchSnapshot();
	});
});

describe("ModalHeader", () => {
	it("Render", () => {
		const { container } = render(
			<BaseTestComponent>
				<Modal.Header />
			</BaseTestComponent>
		);
		expect(container).toMatchSnapshot();
	});
});

describe("ModalFooter", () => {
	it("Render", () => {
		const { container } = render(
			<BaseTestComponent>
				<Modal.Footer />
			</BaseTestComponent>
		);
		expect(container).toMatchSnapshot();
	});
});

describe("ModalCancelButton", () => {
	it("Render", () => {
		const { container } = render(
			<BaseTestComponent>
				<Modal.CancelButton />
			</BaseTestComponent>
		);
		expect(container).toMatchSnapshot();
	});

	it("onClick", async () => {
		const onClose = vi.fn();
		const onCancel = vi.fn();

		render(
			<BaseTestComponent>
				<Modal isOpen onClose={onClose} onCancel={onCancel}>
					<Modal.CancelButton />
				</Modal>
			</BaseTestComponent>
		);

		const button = screen.getByRole("button", { name: /Cancel/i });
		await userEvent.click(button);

		expect(onClose).toBeCalledTimes(1);
		expect(onCancel).toBeCalledTimes(1);
	});
});

describe("ModalAcceptButton", () => {
	it("Render", () => {
		const { container } = render(
			<BaseTestComponent>
				<Modal.AcceptButton />
			</BaseTestComponent>
		);
		expect(container).toMatchSnapshot();
	});

	it("onClick", async () => {
		const onClose = vi.fn();
		const onAccept = vi.fn();

		render(
			<BaseTestComponent>
				<Modal isOpen onClose={onClose} onAccept={onAccept}>
					<Modal.AcceptButton />
				</Modal>
			</BaseTestComponent>
		);

		const button = screen.getByRole("button", { name: /Accept/i });
		await userEvent.click(button);

		expect(onClose).toBeCalledTimes(1);
		expect(onAccept).toBeCalledTimes(1);
	});
});

describe("ModalCloseButton", () => {
	it("Render", () => {
		const { container } = render(
			<BaseTestComponent>
				<Modal.CloseButton />
			</BaseTestComponent>
		);
		expect(container).toMatchSnapshot();
	});
});
