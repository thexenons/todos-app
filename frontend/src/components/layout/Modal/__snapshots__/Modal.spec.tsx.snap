// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal > Render 1`] = `<div />`;

exports[`ModalAcceptButton > Render 1`] = `
<div>
  <button
    class="styled__ButtonWrapper-sc-f8cd241f-0 bYSNlB"
  >
    Accept
  </button>
</div>
`;

exports[`ModalCancelButton > Render 1`] = `
<div>
  <button
    class="styled__ButtonWrapper-sc-f8cd241f-0 bSHSQa"
  >
    Cancel
  </button>
</div>
`;

exports[`ModalCloseButton > Render 1`] = `
<div>
  <button
    class="styled__ButtonWrapper-sc-f8cd241f-0 dEVFMN styled__ModalCloseButton-sc-5dfa476c-5 dvBUPH"
  >
    <span
      class="styled__ButtonIconsWrapper-sc-f8cd241f-1 ctjBQt"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </span>
  </button>
</div>
`;
